File: src/gui/mainwindow.cpp
/*
          __   _,--="=--,_   __
         /  \."    .-.    "./  \
        /  ,/  _   : :   _  \/` \
        \  `| /o\  :_:  /o\ |\__/
         `-'| :="~` _ `~"=: |
            \`     (_)     `/
     .-"-.   \      |      /   .-"-.
.---{     }--|  /,.-'-.,\  |--{     }---.
 )  (_)_)_)  \_/`~-===-~`\_/  (_(_(_)  (
(      Coding like a dog @cycleke       )
 )                                     (
'---------------------------------------'
  */

#include <QAction>
#include <QMenuBar>
#include <QMessageBox>
#include <QStatusBar>

#include "mainwindow.h"
#include "ui_aboutFaceRecognition.h"
#include "ui_mainwindow.h"

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), ui(new Ui::MainWindow) {

  ui->setupUi(this);
  this->setWindowTitle(tr("Face Recognition"));

  // Add MenuBar Action
  connect(ui->actionPreferencs, &QAction::triggered, this,
          &MainWindow::openPreferences);
  connect(ui->actionAbout_Face_Recognition, &QAction::triggered, this,
          &MainWindow::openAboutFaceRecognition);
  connect(ui->actionAbout_Qt, &QAction::triggered, this,
          &MainWindow::openAboutQt);

  // Add PushButton Action

  // Init
  ui->pushButton_OpenCamera->setDisabled(false);
  ui->pushButton_CloseCamera->setDisabled(true);
  ui->pushButton_AddNewFaces->setDisabled(true);
  ui->pushButton_DeleteFaces->setDisabled(true);
}

MainWindow::~MainWindow() { delete ui; }

void MainWindow::openPreferences() {}

void MainWindow::openAboutFaceRecognition() {
  QDialog *dialog = new QDialog(this);
  Ui::AboutFaceRecognition *ui = new Ui::AboutFaceRecognition;

  ui->setupUi(dialog);
  dialog->setWindowTitle(tr("About Face Recognition"));
  dialog->exec();

  delete ui;
  delete dialog;
}

void MainWindow::openAboutQt() { QMessageBox::aboutQt(this); }

void MainWindow::on_pushButton_OpenCamera_clicked() {
  ui->pushButton_OpenCamera->setDisabled(true);
  ui->pushButton_CloseCamera->setDisabled(false);
  ui->pushButton_AddNewFaces->setDisabled(false);
  ui->pushButton_DeleteFaces->setDisabled(false);
}

void MainWindow::on_pushButton_CloseCamera_clicked() {
  ui->pushButton_OpenCamera->setDisabled(false);
  ui->pushButton_CloseCamera->setDisabled(true);
  ui->pushButton_AddNewFaces->setDisabled(true);
  ui->pushButton_DeleteFaces->setDisabled(true);
}

void MainWindow::on_pushButton_AddNewFaces_clicked() {}

void MainWindow::on_pushButton_DeleteFaces_clicked() {}

File: utils/ui/mainwindow.py
# -*- coding: utf-8 -*-

import cv2
from PyQt5.QtCore import QTimer
from PyQt5.QtGui import QPixmap, QImage
from PyQt5.QtWidgets import QMainWindow, QMessageBox, QDialog

from utils.core.recognizer import Recognizer
from utils.ui.ui_aboutFaceRecognition import Ui_AboutFaceRecognition
from utils.ui.ui_getName import Ui_getName
from utils.ui.ui_mainwindow import Ui_MainWindow


class MainWindow(QMainWindow):

    def __init__(self, parent=None):
        QMainWindow.__init__(self, parent=parent)

        self.ui = Ui_MainWindow()
        self.ui.setupUi(self)

        self.ui.label_CameraShow.setScaledContents(True)

        self.timer = QTimer(self)
        self.timer.timeout.connect(self.recognize_face)

        self.capture = cv2.VideoCapture()
        self.capture.release()

        self.recognizer = Recognizer()
        self.recognizer.load_data()

        self.ui.actionPreferencs.triggered.connect(self.open_preferences)
        self.ui.actionAbout_Face_Recognition.triggered.connect(self.open_about_face_recognition)
        self.ui.actionAbout_Qt.triggered.connect(self.open_about_qt)

        self.ui.pushButton_OpenCamera.clicked.connect(self.clicked_open_camera)
        self.ui.pushButton_CloseCamera.clicked.connect(self.clicked_close_camera)
        self.ui.pushButton_AddNewFaces.clicked.connect(self.clicked_add_new_faces)
        self.ui.pushButton_DeleteFaces.clicked.connect(self.clicked_delete_faces)

        self.ui.pushButton_OpenCamera.setDisabled(False)
        self.ui.pushButton_CloseCamera.setDisabled(True)
        self.ui.pushButton_AddNewFaces.setDisabled(False)
        self.ui.pushButton_DeleteFaces.setDisabled(False)

    def open_preferences(self):
        pass

    def open_about_face_recognition(self):
        dialog = QDialog(parent=self)
        ui = Ui_AboutFaceRecognition()
        ui.setupUi(dialog)
        dialog.setWindowTitle("About Face Recognition")
        dialog.exec()

    def open_about_qt(self):
        QMessageBox.aboutQt(self)

    def clicked_open_camera(self):
        self.ui.pushButton_OpenCamera.setDisabled(True)
        self.ui.pushButton_CloseCamera.setDisabled(False)

        if self.capture.isOpened():
            self.capture.release()

        self.capture.open(0)
        if self.capture.isOpened():
            rate = self.capture.get(cv2.CAP_PROP_FPS)

            success, frame = self.capture.read()
            if frame.any():
                img = self.recognizer.recognise_image(frame)
                cv2.cvtColor(img, cv2.COLOR_BGR2RGB, img)
                height, width, bytes_per_component = img.shape
                bytes_per_line = 3 * width
                q_img = QImage(img.data, width, height, bytes_per_line, QImage.Format_RGB888)
                pixel_map = QPixmap.fromImage(q_img)
                self.ui.label_CameraShow.setPixmap(pixel_map)
                self.timer.setInterval(1000 / rate)
                self.timer.start()

    def recognize_face(self):
        success, frame = self.capture.read()
        img = self.recognizer.recognise_image(frame)
        cv2.cvtColor(img, cv2.COLOR_BGR2RGB, img)
        height, width, bytes_per_component = img.shape
        bytes_per_line = 3 * width
        q_img = QImage(img.data, width, height, bytes_per_line, QImage.Format_RGB888)
        pixel_map = QPixmap.fromImage(q_img)
        self.ui.label_CameraShow.setPixmap(pixel_map)

    def clicked_close_camera(self):
        self.ui.pushButton_OpenCamera.setDisabled(False)
        self.ui.pushButton_CloseCamera.setDisabled(True)

        self.timer.stop()
        self.capture.release()
        self.ui.label_CameraShow.setPixmap(QPixmap())

    def clicked_add_new_faces(self):
        dialog = QDialog(self)
        ui = Ui_getName()
        ui.setupUi(dialog)
        dialog.setWindowTitle("Add New Faces")

        if dialog.exec():
            name = ui.lineEdit.text()
            if name:
                self.recognizer.add_face(name)

    def clicked_delete_faces(self):
        dialog = QDialog(self)
        ui = Ui_getName()
        ui.setupUi(dialog)
        dialog.setWindowTitle("Delete Faces")

        if dialog.exec():
            name = ui.lineEdit.text()
            if name in self.recognizer.face_descriptors.keys():
                self.recognizer.face_descriptors.pop(name)
